controllers/metal3.io/baremetalhost_controller.go

```diff
@@ -1153,6 +1180,245 @@ func saveHostProvisioningSettings(host *metal3v1alpha1.BareMetalHost) (dirty boo
+	}
+	if !k8serrors.IsNotFound(err) {
+		// Error reading the object
+		return errors.Wrap(err, "could not load host firmware settings")
+	}
+
+	// Get the current settings and schema
+	currentSettings, schema, err := prov.GetFirmwareSettings(true)
+	if err != nil {
+		info.log.Info("could not get firmware settings from provisioner")
+		return err
+	}
+
+	// Create a new resource
+	if err = r.newHostFirmwareSettings(info, hfs); err != nil {
+		return errors.Wrap(err, "could not create hostFirmwareSettings resource")
+	}
+
+	if schema != nil && len(schema) > 0 {
+
+		firmwareSchema, err := r.getOrCreateFirmwareSchema(info, hfs, schema)
+		if err != nil {
+			return errors.Wrap(err, "could not get/create firmware schema in cluster")
+		}
+
+		// Set hostFirmwareSetting to use this schema
+		hfs.Status.FirmwareSchema = &metal3v1alpha1.SchemaReference{
+			Namespace: firmwareSchema.ObjectMeta.Namespace,
+			Name:      firmwareSchema.ObjectMeta.Name}
+
+		// copy schema reference to Annotation so it can be restored after pivoting
+		hfs.Annotations[metal3v1alpha1.FirmwareSchemaNamespaceAnnotation] = hfs.Status.FirmwareSchema.Namespace
+		hfs.Annotations[metal3v1alpha1.FirmwareSchemaNameAnnotation] = hfs.Status.FirmwareSchema.Name
+
+		if err = r.updateHostFirmwareSettingsFromCurrent(currentSettings, hfs, firmwareSchema); err != nil {
+			return errors.Wrap(err, "could not update hostFirmwareSettings")
+		}
+	} else {
+		if err = r.updateHostFirmwareSettingsFromCurrent(currentSettings, hfs, nil); err != nil {
+			return errors.Wrap(err, "could not update hostFirmwareSettings")
+		}
+		info.log.Info("updated hostFirmwareSettings settings without schema")
+	}
+
+	return nil
+}
+
+// Create a HostFirmwareSettings resource
+func (r *BareMetalHostReconciler) newHostFirmwareSettings(info *reconcileInfo, hfs *metal3v1alpha1.HostFirmwareSettings) (err error) {
+
+	// Use same name and namespace as host
+	hfs.ObjectMeta = metav1.ObjectMeta{
+		Name:      info.host.ObjectMeta.Name,
+		Namespace: info.host.ObjectMeta.Namespace}
+	hfs.Status.Settings = make(metal3v1alpha1.SettingsMap)
+	hfs.Spec.Settings = make(metal3v1alpha1.DesiredSettingsMap)
+	hfs.Annotations = make(map[string]string)
+	if err = r.Create(context.TODO(), hfs); err != nil {
+		return errors.Wrap(err, "could not create hostFirmwareSettings resource")
+	}
+	info.log.Info("created new hostFirmwareSettings resource")
+
+	return nil
+}
+
+// Update the HostFirmwareSettings resource using the settings and schema from provisioner
+func (r *BareMetalHostReconciler) updateHostFirmwareSettingsFromCurrent(settings metal3v1alpha1.SettingsMap, hfs *metal3v1alpha1.HostFirmwareSettings, schema *metal3v1alpha1.FirmwareSchema) (err error) {
+
+	// if no schema provided, get the one referenced by hostFirmwareSettings if available
+	if schema == nil && hfs.Status.FirmwareSchema != nil {
+		schema = &metal3v1alpha1.FirmwareSchema{}
+		if r.Get(context.TODO(), client.ObjectKey{
```

Comment:
I have a feeling that this is redundant since all places where updateHostFirmwareSettingsFromCurrent is called already try to fetch the current schema.

Reply:
Yes, good catch. Removed this.